README.md
# 🔥 Nuxt Flame

[![npm version][npm-version-src]][npm-version-href]
[![npm downloads][npm-downloads-src]][npm-downloads-href]
[![License][license-src]][license-href]

Easily integrate Google Firebase into your Nuxt 3 application.

- [✨ &nbsp;Release Notes](/CHANGELOG.md)
- [🏗️ &nbsp;Installation](#installation)
- [⚙️ &nbsp;Usage](#usage)
- [🦾 &nbsp;Advanced Usage](#advanced-usage)

## Features

- ⛰ &nbsp;Firebase Auth, Realtime Database, Firestore, Cloud Functions and Cloud Storage
- 🏎️ &nbsp;Full emulators support
- 🔋 &nbsp;SSR Friendly
- 🔐 &nbsp;Safe
- 🪶 &nbsp;Extremely light

## Installation

1. Add `nuxt-flame` dependency to your project
```bash
# Using pnpm
pnpm add -D nuxt-flame firebase firebase-admin

# Using yarn
yarn add --dev nuxt-flame firebase firebase-admin

# Using npm
npm install --save-dev nuxt-flame firebase firebase-admin
```

2. Add `nuxt-flame` to the `modules` section of `nuxt.config.ts` and specify Firebase credentials in `runtimeConfig`
```js
export default defineNuxtConfig({
  modules: [
    "nuxt-flame"
  ],

  runtimeConfig: {
    firebaseAdminCredentials: {
      projectId: process.env.FIREBASE_PROJECT_ID,
      clientEmail: process.env.FIREBASE_CLIENT_EMAIL,
      privateKey: process.env.FIREBASE_PRIVATE_KEY,
    },

    public: {
      firebaseCredentials: {
        apiKey: process.env.FIREBASE_API_KEY,
        authDomain: process.env.FIREBASE_AUTH_DOMAIN,
        projectId: process.env.FIREBASE_PROJECT_ID,
        storageBucket: process.env.FIREBASE_STORAGE_BUCKET,
        messagingSenderId: process.env.FIREBASE_MESSAGING_SENDER_ID,
        appId: process.env.FIREBASE_APP_ID,
        measurementId: process.env.FIREBASE_MEASUREMENT_ID,
      },
    },
  },
})
```

3. Create `.env` file and put your credentials there
```js
// Web credentials from code snippet provided by Google Firebase
FIREBASE_API_KEY=********
FIREBASE_AUTH_DOMAIN=********
FIREBASE_PROJECT_ID=********
FIREBASE_STORAGE_BUCKET=********
FIREBASE_MESSAGING_SENDER_ID=********
FIREBASE_APP_ID=********
FIREBASE_MEASUREMENT_ID=********

// Admin credentials from service account key
FIREBASE_CLIENT_EMAIL=********
FIREBASE_PRIVATE_KEY="-----BEGIN PRIVATE KEY-----
****************************************************
****************************************************
****************************************************
****************************************************
****************************************************
****************************************************
**************************
-----END PRIVATE KEY-----"
```

4. _(Optional)_ Feel free to configure Nuxt Flame in `nuxt.config.ts`
```js
export default defineNuxtConfig({
  // ...

  /**
   * Default Nuxt Flame settings
   */
  flame: {
    authApiEndpoint: "/api/__session",
    authCookieName: "__session",

    emulators: {
      enabled: false,

      auth: {
        enabled: false,
        url: "http://127.0.0.1:9099",
        options: {
          disableWarnings: false,
        },
      },

      firestore: {
        enabled: false,
        host: "127.0.0.1",
        port: 8080,
      },

      database: {
        enabled: false,
        host: "127.0.0.1",
        port: 9000,
      },

      functions: {
        enabled: false,
        host: "127.0.0.1",
        port: 5001,
      },

      storage: {
        enabled: false,
        host: "127.0.0.1",
        port: 9199,
      },
    },
  },

  // ...
})
```

That's it! You can now use Nuxt Flame in your Nuxt app ✨

## Usage

Nuxt Flame exports bunch of Firebase helpers (Vue composables) that available both client and server side.


Get Firebase App **(client-side only)**:
```js
const app = useFirebaseApp()
```

Get Firebase Admin App **(server-side only)**:
```js
const admin = useFirebaseAdminApp()
```

Get Firebase Auth instance **(client-side only)**:
```js
const auth = useAuth()
```

Get Firebase Auth instance **(server-side only)**:
```js
const auth = useServerAuth()
```

Get current user. **Important:** Client returns `User` object when server returns `DecodedIdToken` object:
```js
const currentUser = useCurrentUser()
```

Get Firebase Realtime Database instance:
```js
const db = useDatabase()
```

Get Firebase Firestore instance:
```js
const db = useFirestore()
```

Get Firebaes Cloud Functions instance:
```js
const functions = useFunctions()
```

Get Firebaes Cloud Storage instance:
```js
const storage = useStorage()
```

## Advanced Usage

TBD

## Roadmap

- [ ] Write tests
- [ ] Create helpers for Firestore
- [ ] Create helpers for Functions
- [ ] Create helpers for Storage
- [ ] Create helpers for Realtime Database

## Development

```bash
# Install dependencies
pnpm install

# Generate type stubs
pnpm run dev:prepare

# Develop with the playground
pnpm run dev

# Build the playground
pnpm run dev:build

# Run ESLint
pnpm run lint

# Run Vitest
pnpm run test
pnpm run test:watch

# Release new version
pnpm run release
```

[npm-version-src]: https://img.shields.io/npm/v/nuxt-flame/latest.svg?style=flat&colorA=18181B&colorB=28CF8D
[npm-version-href]: https://npmjs.com/package/nuxt-flame

[npm-downloads-src]: https://img.shields.io/npm/dm/nuxt-flame.svg?style=flat&colorA=18181B&colorB=28CF8D
[npm-downloads-href]: https://npmjs.com/package/nuxt-flame

[license-src]: https://img.shields.io/npm/l/nuxt-flame.svg?style=flat&colorA=18181B&colorB=28CF8D
[license-href]: https://npmjs.com/package/nuxt-flame


CHANGELOG.md
## v1.0.0


### 🔥 Initial Release

### ❤️  Contributors

- Andrew Kodkod ([@AKodkod](http://github.com/AKodkod))

src/module.ts
import { defineNuxtModule, addPlugin, createResolver, addServerHandler, addImports } from "@nuxt/kit"
import { AuthEmulatorConfig, CommonEmulatorConfig } from "./runtime/utils/emulators"

/**
 * Nuxt module options
 */
export interface NuxtFlameOptions {
  /**
   * The name of the key to use to save auth state
   * @default "nuxtFlameCurrentUser"
   * @example "currentUser"
   */
  authStateKey?: string

  /**
   * The endpoint to use for the auth API
   * @default "/api/__session"
   * @example "/api/auth"
   */
  authApiEndpoint?: string

  /**
   * The name of the cookie to use for the auth API
   * @default "__session"
   * @example "auth"
   */
  authCookieName?: string

  /**
   * Emulators configuration
   */
  emulators: {
    /**
     * Enable all emulators. This will override all other emulator options.
     * @default false
     */
    enabled?: boolean

    /**
     * Auth emulator configuration
     */
    auth?: Partial<AuthEmulatorConfig>

    /**
     * Firestore emulator configuration
     */
    firestore?: Partial<CommonEmulatorConfig>

    /**
     * Realtime Database emulator configuration
     */
    database?: Partial<CommonEmulatorConfig>

    /**
     * Cloud Functions emulator configuration
     */
    functions?: Partial<CommonEmulatorConfig>

    /**
     * Cloud Storage emulator configuration
     */
    storage?: Partial<CommonEmulatorConfig>
  }
}

export interface NuxtFlameOptionsFull {
  authStateKey: string
  authApiEndpoint: string
  authCookieName: string

  emulators: {
    enabled: boolean
    auth: AuthEmulatorConfig,
    firestore: CommonEmulatorConfig,
    database: CommonEmulatorConfig,
    functions: CommonEmulatorConfig,
    storage: CommonEmulatorConfig,
  }
}

export default defineNuxtModule<NuxtFlameOptions>({
  meta: {
    name: "nuxt-flame",
    configKey: "flame",
    compatibility: {
      nuxt: "^3.0.0",
    },
  },

  /**
   * Default options
   */
  defaults: {
    authStateKey: "nuxtFlameCurrentUser",
    authApiEndpoint: "/api/__session",
    authCookieName: "__session",

    emulators: {
      enabled: false,

      auth: {
        enabled: false,
        url: "http://127.0.0.1:9099",
        options: {
          disableWarnings: false,
        },
      },

      firestore: {
        enabled: false,
        host: "127.0.0.1",
        port: 8080,
      },

      database: {
        enabled: false,
        host: "127.0.0.1",
        port: 9000,
      },

      functions: {
        enabled: false,
        host: "127.0.0.1",
        port: 5001,
      },

      storage: {
        enabled: false,
        host: "127.0.0.1",
        port: 9199,
      },
    },
  },

  setup(options, nuxt) {
    // Save the options to the Nuxt runtime config
    nuxt.options.runtimeConfig.public.flame = options as NuxtFlameOptionsFull

    // Resolve the module directory
    const resolver = createResolver(import.meta.url)

    // Load client and server plugins
    addPlugin(resolver.resolve("./runtime/plugin.client"))
    addPlugin(resolver.resolve("./runtime/plugin.server"))

    // Add the auth API endpoint
    addServerHandler({
      route: options.authApiEndpoint,
      handler: resolver.resolve("./runtime/server/api/session.server"),
    })

    // Import the composable functions
    addImports([
      {
        from: resolver.resolve("./runtime/composables/use-flame-config"),
        name: "useFlameConfig",
      },
      {
        from: resolver.resolve("./runtime/composables/use-firebase-app.client"),
        name: "useFirebaseApp",
      },
      {
        from: resolver.resolve("./runtime/composables/use-firebase-admin-app.server"),
        name: "useFirebaseAdminApp",
      },
      {
        from: resolver.resolve("./runtime/composables/use-auth"),
        name: "useAuth",
      },
      {
        from: resolver.resolve("./runtime/composables/use-server-auth.server"),
        name: "useServerAuth",
      },
      {
        from: resolver.resolve("./runtime/composables/use-current-user"),
        name: "useCurrentUser",
      },
      {
        from: resolver.resolve("./runtime/composables/use-firestore"),
        name: "useFirestore",
      },
      {
        from: resolver.resolve("./runtime/composables/use-collection"),
        name: "useCollection",
      },
      {
        from: resolver.resolve("./runtime/composables/use-async-collection"),
        name: "useAsyncCollection",
      },
      {
        from: resolver.resolve("./runtime/composables/use-document"),
        name: "useDocument",
      },
      {
        from: resolver.resolve("./runtime/composables/use-async-document"),
        name: "useAsyncDocument",
      },
      {
        from: resolver.resolve("./runtime/composables/use-document-subscribe"),
        name: "useDocumentSubscribe",
      },

      {
        from: resolver.resolve("./runtime/composables/use-collection-subscribe"),
        name: "useCollectionSubscribe",
      },
      {
        from: resolver.resolve("./runtime/composables/use-database"),
        name: "useDatabase",
      },
      {
        from: resolver.resolve("./runtime/composables/use-functions"),
        name: "useFunctions",
      },
      {
        from: resolver.resolve("./runtime/composables/use-function"),
        name: "useFunction",
      },
      {
        from: resolver.resolve("./runtime/composables/use-storage"),
        name: "useStorage",
      },
    ])
  },
})
